witness/golang/omniwitness/usbarmory/internal/storage/slots/journal.go

```diff
@@ -0,0 +1,295 @@
+// [start, start+length) range of blocks accessible via dev.
+// Journal ranges should not overlap with one another, or corruption will almost certainly occur.
+func OpenJournal(dev BlockReaderWriter, start, length uint) (*Journal, error) {
+	j := &Journal{
+		dev:          dev,
+		start:        start,
+		length:       length,
+		maxDataBytes: (length*dev.BlockSize())/minEntries - entryHeaderSize,
+	}
+
+	if err := j.init(); err != nil {
+		return nil, err
+	}
+
+	return j, nil
+}
+
+// Data returns the application data from the most recent valid entry in the
+// journal, along with the entry's revision number.
+// If the returned revision is zero, then no successful writes have taken place
+// on this journal.
+func (j *Journal) Data() ([]byte, uint32) {
+	return j.current.Data, j.current.Revision
+}
+
+// Update creates a new entry record in the journal for the provided data.
+// The new record's revision will be one greater than the previous record (or 1
+// if no previous record exists).
+func (j *Journal) Update(data []byte) error {
+	if l := len(data); l > int(j.maxDataBytes) {
+		return fmt.Errorf("attemping to write %d bytes, larger than the max permitted in this journal (%d bytes)", l, j.maxDataBytes)
+	}
+	e := &entry{
+		Magic:      [4]byte{magic0[0], magic0[1], magic0[2], magic0[3]},
+		Revision:   j.current.Revision + 1,
+		DataLen:    uint64(len(data)),
+		DataSHA256: sha256.Sum256(data),
+		Data:       data,
+	}
+
+	cap := (j.start + j.length - j.nextBlock) * j.dev.BlockSize()
+	if cap < uint(e.Size()) {
+		// The record won't fit in the remaining space, so wrap around and write at the beginning.
+		j.nextBlock = j.start
+	}
+	// TODO(al): consider making this more "streamy".
+	buf := &bytes.Buffer{}
+	if err := marshalEntry(e, buf); err != nil {
+		return fmt.Errorf("failed to marshal entry: %v", err)
+	}
+	return j.dev.WriteBlocks(j.nextBlock, buf.Bytes())
+}
+
+// Init scans the journal to figure out the latest valid record, if any.
+func (j *Journal) init() error {
+	// Start where all good stories do: at the beginning!
+	lba := j.start
+	var lastEntry entry
+	nextWriteLBA := j.start
+	for lba < j.start+j.length {
+		br := newBlockReader(j.dev, lba)
+		e, err := unmarshalEntry(br)
+		if err != nil {
+			if lastEntry.Revision > 0 {
+				// We already found the lastet record in the journal, so we're done.
+				break
+			}
+			// Since we haven't already found a good entry, and we were unable
+			// to unmarshal this one,
+			// this means that either:
+			//  a) the journal is completely empty, or
+			//  b) the previously good entry/ies at the start of the journal
+			//     have been completely or partially overwritten during a
+			//     failed write attempt.
+			// Either way, we don't have a valid entry wth a length field we can
+			// rely on, so we we'll have to fall back to scanning all blocks to
+			// look for one.
+			lba++
+			continue
+		}
+		if lastEntry.Revision == 0 || e.Revision > lastEntry.Revision {
```

Comment:
I suspect `lastEntry.Revision == 0` is redundant, because if it is 0 then the next check should trivially pass (because revision is uint, and real revisions are always positive).

Reply:
Yeah, it was a mechanical change to replace the `nil` check with `Rev==0`, as you say there is a material difference in that previously a written entry which had a Revision of 0 would have been accepted, but in practice that should never happen so we can safely simplify this.